presto-hive/src/main/java/com/facebook/presto/hive/s3/PrestoS3FileSystem.java

```diff
@@ -1088,4 +1088,18 @@ void setS3Client(AmazonS3 client)
     {
         s3 = client;
     }
+
+    /**
+     * Helper function used to work around the fact that if you use an S3 bucket with an '_' that java.net.URI
+     * does behaves differently and sets the host value to null where as S3 buckets without '_' have a properly
+     * set host field.
+     *
+     * @param uri The URI from which to extract a host value.
+     * @return The host value where uri.getAuthority() is used when uri.getHost() returns null.
+     * @note '_' is only allowed in S3 bucket names in us-east-1
+     */
+    String safeGetHost(URI uri)
+    {
+        return uri.getHost() != null ? uri.getHost() : uri.getAuthority();
```

Comment:
This seems potentially dangerous since authority can contain the username/password. What if we only perform this fallback if `isPathStyleAccess` is true, and throw an exception otherwise (since the URI is invalid)?

Additionally, we could call `uri.getUserInfo()` and fail if it returns non-null, since that would mean authority does contain a username/password.

Reply:
I think there may be bigger security issues if the URI contains credentials in that fashion (wouldn't that imply basic auth?) but I'll confirm this behavior and change it accordingly since it doesn't cost anything to be a bit more defensive here.

Reply:
yeah it's not secure, but I have seen people use that.

Reply:
It still uses normal S3 auth. They get pulled out in `PrestoS3FileSystem.getAwsCredentials()`. I completely agree that this is a bad idea, but we added it to be compatible with existing Hadoop S3 setups.

Reply:
I've made all the changes you suggested except the pathStyleAccess check. Based on the documentation of pathStyleAccess in the S3 client referring to more than just the bucket name and the potential for that setting to be erroneous false in relation to the bucket name I felt it was more accurate to use the results of URI inspection. Let me know if you feel strongly but this latest update likely has fewer edge-cases in my opinion.